README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

<img src="https://raw.githubusercontent.com/matt-dray/stickers/master/output/dialga_hex.png" width="150" align="right">

# {dialga}

<!-- badges: start -->

[![Project Status: Inactive – The project has reached a stable, usable
state but is no longer being actively developed; support/maintenance
will be provided as time
allows.](https://www.repostatus.org/badges/latest/inactive.svg)](https://www.repostatus.org/#inactive)
[![R-CMD-check](https://github.com/matt-dray/dialga/workflows/R-CMD-check/badge.svg)](https://github.com/matt-dray/dialga/actions)
[![Codecov test
coverage](https://codecov.io/gh/matt-dray/dialga/branch/main/graph/badge.svg)](https://codecov.io/gh/matt-dray/dialga?branch=main)
[![CodeFactor](https://www.codefactor.io/repository/github/matt-dray/dialga/badge)](https://www.codefactor.io/repository/github/matt-dray/dialga)
[![rostrum.blog
post](https://img.shields.io/badge/rostrum.blog-post-008900?style=flat&labelColor=black&logo=data:image/gif;base64,R0lGODlhEAAQAPEAAAAAABWCBAAAAAAAACH5BAlkAAIAIf8LTkVUU0NBUEUyLjADAQAAACwAAAAAEAAQAAAC55QkISIiEoQQQgghRBBCiCAIgiAIgiAIQiAIgSAIgiAIQiAIgRAEQiAQBAQCgUAQEAQEgYAgIAgIBAKBQBAQCAKBQEAgCAgEAoFAIAgEBAKBIBAQCAQCgUAgEAgCgUBAICAgICAgIBAgEBAgEBAgEBAgECAgICAgECAQIBAQIBAgECAgICAgICAgECAQECAQICAgICAgICAgEBAgEBAgEBAgICAgICAgECAQIBAQIBAgECAgICAgIBAgECAQECAQIBAgICAgIBAgIBAgEBAgECAgECAgICAgICAgECAgECAgQIAAAQIKAAAh+QQJZAACACwAAAAAEAAQAAAC55QkIiESIoQQQgghhAhCBCEIgiAIgiAIQiAIgSAIgiAIQiAIgRAEQiAQBAQCgUAQEAQEgYAgIAgIBAKBQBAQCAKBQEAgCAgEAoFAIAgEBAKBIBAQCAQCgUAgEAgCgUBAICAgICAgIBAgEBAgEBAgEBAgECAgICAgECAQIBAQIBAgECAgICAgICAgECAQECAQICAgICAgICAgEBAgEBAgEBAgICAgICAgECAQIBAQIBAgECAgICAgIBAgECAQECAQIBAgICAgIBAgIBAgEBAgECAgECAgICAgICAgECAgECAgQIAAAQIKAAA7)](https://www.rostrum.blog/2021/04/10/dialga/)
<!-- badges: end -->

## Purpose

{dialga} is an R package that lets you build and interpret standard
[cron expressions](https://en.wikipedia.org/wiki/Cron#CRON_expression)
using familiar R syntax.

Cron is software for scheduling computer tasks (‘cron jobs’). Cron
strings detail concisely the required schedule. They require a specific
format, like `"0/15 * 1,3,20 6 0,6"`, but it can be difficult to
remember how to structure them.

Read more in [the accompanying blog
post](https://www.rostrum.blog/2021/04/10/dialga/) and [report any
bugs](https://github.com/matt-dray/dialga/issues) you might find.

## Install

Install from GitHub with help from {remotes}:

``` r
install.packages("remotes")  # if not already installed
remotes::install_github("matt-dray/dialga")
library(dialga)
```

If you want to be able to have the cron string output copied directly to
your clipboard, you’ll need to install [the {clipr}
package](https://cran.r-project.org/package=clipr) to your machine with
`install.packages("clipr")` as well. You don’t have to install it if you
don’t want to.

## Demonstration

There’s currently two functions: `r2cron()` takes integer vectors as
inputs to time-period arguments and spits out a cron string, and
`cron2eng()` takes a valid cron string and prints out a readable English
version. See `?dialga::r2cron()` and `?dialga::cron2eng()` for further
details.

### Simple

A simple example of `r2cron()`: how would you specify the 28th minute
past 11PM every day?

``` r
x <- dialga::r2cron(
  minutes = 28, 
  hours = 23  # 24-hour clock
)

x
#> [1] "28 23 * * *"
```

Note that you can use the argument `clip = TRUE` to have the output
copied to your system’s clipboard so you can paste it elsewhere. You
must install [the {clipr}
package](https://cran.r-project.org/package=clipr) separately if you
want this functionality.

To interpret the output—and confirm we got what we expected—we can pass
that cron string into `cron2eng()`.

``` r
dialga::cron2eng(x)
#> Cron string '28 23 * * *' means:
#>   - minute(s) 28
#>   - hour(s) 11PM
#>   - every day(s) of the month
#>   - every month(s)
#>   - any day(s) of the week
```

The output isn’t sophisticated, but it communicates the point.

You could even pipe these functions together to go from R to English.

``` r
library(magrittr)  # for %>%

dialga::r2cron(minutes = 28, hours = 23) %>% 
  dialga::cron2eng()
#> Cron string '28 23 * * *' means:
#>   - minute(s) 28
#>   - hour(s) 11PM
#>   - every day(s) of the month
#>   - every month(s)
#>   - any day(s) of the week
```

### Complex

A more complicated (i.e. contrived) request might be ‘every 20 minutes
from the zeroth minute of 3PM, 4PM and 5PM, on the 1st days of April,
October and November, plus every weekend’:

``` r
y <- dialga::r2cron(
 minutes = seq(0, 59, 20),
 hours = 15:17,  # 24-hr clock
 days_month = 1,
 months = c(4, 10, 11),
 days_week = c(1, 7)  # Sunday is '1'
)

y
#> [1] "0/20 15-17 1 4,10,11 0,6"
```

And in English:

``` r
dialga::cron2eng(y)
#> Cron string '0/20 15-17 1 4,10,11 0,6' means:
#>   - every 20 minute(s) starting from minute(s) 0
#>   - hour(s) 3PM to 5PM
#>   - day(s) of the month 1
#>   - month(s) April, October and November
#>   - and day(s) of the week Sunday and Saturday
```

### Warnings

As a courtesy, you’ll be warned when unlikely dates arise:

``` r
dialga::r2cron(days_month = 28:31, months = 2)
#> Warning in dialga::r2cron(days_month = 28:31, months = 2): 
#>   Sure? There's no 31st in Feb, Apr, Jun, Sept nor Nov.
#> Warning in dialga::r2cron(days_month = 28:31, months = 2): 
#>   Sure? There's no 30th in Feb.
#> Warning in dialga::r2cron(days_month = 28:31, months = 2): 
#>   Sure? 29 Feb is only in leap years.
#> [1] "* * 28-31 2 *"
```

## Actual scheduling tools

If on Unix/Linux, you can use [the {cronR}
package](https://github.com/bnosac/cronR) to schedule tasks from R. The
Windows alternative is [the {taskscheduleR}
package](https://github.com/bnosac/taskscheduleR). I typically use the
web service [crontab.guru](https://crontab.guru) to build cron
expressions. It was helpful for checking {dialga}’s functionality.

## Code of Conduct

Please note that the {dialga} project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# dialga 0.1.1

* Refactor `cron2eng()` translation of hours, months, day of week containing `/` (#7)

# dialga 0.1

* Eliminated {knitr} dependency (#11)
* Make `clip` argument `FALSE` by default in `r2cron()` (#12)
* Eliminated {stringr} dependency (#11)
* Add 'and' to specified day of week for clarity
* Eliminated {stringr} dependency (#11)
* Fixed bug where sequences for hours and days-week failed to produce output

# dialga 0.0.0.9002

* Added basic `cron2eng()` function (#5) and improved output format (#10)
* Fixed bug where `days_week = c(1, 7)` was output to `"0/6"` instead of `"0,6"`
* Documented internal functions
* Tested `x/y` edge cases (#9)
* Improved basic test coverage to 100% (#8)

# dialga 0.0.0.9001

* Interpreted `x/y` output format for all time period arguments (#1)
* Changed default inputs to vectors rather than asterisk characters (#2)
* Remove zero-indexing of days of the week (#6)

# dialga 0.0.0.9000

* Added cron.R and utils.R for function `r2cron()`
* Added basic tests for `r2cron()`
* Added package basics, NEWS, COC, license, README
* Added GitHub Actions and badges


tests/testthat.R
library(testthat)
library(dialga)

test_check("dialga")


tests/testthat/test-cron.R

# r2cron() ----------------------------------------------------------------


test_that("empty r2cron() returns all-asterisk string", {
  expect_equal(r2cron(clip = FALSE), "* * * * *")
  expect_length(r2cron(clip = FALSE), 1)
})

test_that("r2cron() minutes argument works", {

 expect_equal(r2cron(minutes = 0:59, clip = FALSE), "* * * * *")
 expect_equal(r2cron(minutes = 1, clip = FALSE), "1 * * * *")
 expect_equal(r2cron(minutes = 1:3, clip = FALSE), "1-3 * * * *")
 expect_equal(r2cron(minutes = 1:2, clip = FALSE), "1-2 * * * *")
 expect_equal(r2cron(minutes = seq(0, 59, 15), clip = FALSE), "0/15 * * * *")
 expect_equal(r2cron(minutes = c(1, 6), clip = FALSE), "1,6 * * * *")
 expect_equal(r2cron(minutes = c(1, 6, 20), clip = FALSE), "1,6,20 * * * *")

 expect_error(r2cron(minutes = "x", clip = FALSE))
 expect_error(r2cron(minutes = TRUE, clip = FALSE))
 expect_error(r2cron(minutes = mtcars, clip = FALSE))
 expect_error(r2cron(minutes = list("x", "y"), clip = FALSE))
 expect_error(r2cron(minutes = 1.2, clip = FALSE))
 expect_error(r2cron(minutes = 60, clip = FALSE))

})

test_that("r2cron() hours argument works", {

 expect_equal(r2cron(hours = 0:23, clip = FALSE), "* * * * *")
 expect_equal(r2cron(hours = 1, clip = FALSE), "* 1 * * *")
 expect_equal(r2cron(hours = 1:3, clip = FALSE), "* 1-3 * * *")
 expect_equal(r2cron(hours = 1:2, clip = FALSE), "* 1-2 * * *")
 expect_equal(r2cron(hours = seq(0, 23, 3), clip = FALSE), "* 0/3 * * *")
 expect_equal(r2cron(hours = c(1, 6), clip = FALSE), "* 1,6 * * *")
 expect_equal(r2cron(hours = c(1, 6, 20), clip = FALSE), "* 1,6,20 * * *")

 expect_error(r2cron(hours = "x", clip = FALSE))
 expect_error(r2cron(hours = TRUE, clip = FALSE))
 expect_error(r2cron(hours = mtcars, clip = FALSE))
 expect_error(r2cron(hours = list("x", "y"), clip = FALSE))
 expect_error(r2cron(hours = 1.2, clip = FALSE))
 expect_error(r2cron(hours = 24, clip = FALSE))

})

test_that("r2cron() days_month argument works", {

 expect_equal(r2cron(days_month = 1:31, clip = FALSE), "* * * * *")
 expect_equal(r2cron(days_month = 1, clip = FALSE), "* * 1 * *")
 expect_equal(r2cron(days_month = 1:3, clip = FALSE), "* * 1-3 * *")
 expect_equal(r2cron(days_month = 1:2, clip = FALSE), "* * 1-2 * *")
 expect_equal(r2cron(days_month = seq(1, 31, 3), clip = FALSE), "* * 1/3 * *")
 expect_equal(r2cron(days_month = c(1, 6), clip = FALSE), "* * 1,6 * *")
 expect_equal(r2cron(days_month = c(1, 6, 20), clip = FALSE), "* * 1,6,20 * *")

 expect_error(r2cron(days_month = "x", clip = FALSE))
 expect_error(r2cron(days_month = TRUE, clip = FALSE))
 expect_error(r2cron(days_month = mtcars, clip = FALSE))
 expect_error(r2cron(days_month = list("x", "y"), clip = FALSE))
 expect_error(r2cron(days_month = 1.2, clip = FALSE))
 expect_error(r2cron(days_month = 32, clip = FALSE))

})

test_that("r2cron() months argument works", {

 expect_equal(r2cron(months = 1:12, clip = FALSE), "* * * * *")
 expect_equal(r2cron(months = 1, clip = FALSE), "* * * 1 *")
 expect_equal(r2cron(months = 1:3, clip = FALSE), "* * * 1-3 *")
 expect_equal(r2cron(months = 1:2, clip = FALSE), "* * * 1-2 *")
 expect_equal(r2cron(months = seq(1, 12, 3), clip = FALSE), "* * * 1/3 *")
 expect_equal(r2cron(months = c(1, 6), clip = FALSE), "* * * 1,6 *")
 expect_equal(r2cron(months = c(1, 6, 7), clip = FALSE), "* * * 1,6,7 *")

 expect_error(r2cron(months = "x", clip = FALSE))
 expect_error(r2cron(months = TRUE, clip = FALSE))
 expect_error(r2cron(months = mtcars, clip = FALSE))
 expect_error(r2cron(months = list("x", "y"), clip = FALSE))
 expect_error(r2cron(months = 1.2, clip = FALSE))
 expect_error(r2cron(months = 13, clip = FALSE))

})

test_that("r2cron() days_week argument works", {

 expect_equal(r2cron(days_week = 1:7, clip = FALSE), "* * * * *")
 expect_equal(r2cron(days_week = 2, clip = FALSE), "* * * * 1")
 expect_equal(r2cron(days_week = 2:4, clip = FALSE), "* * * * 1-3")
 expect_equal(r2cron(days_week = 2:3, clip = FALSE), "* * * * 1-2")
 expect_equal(r2cron(days_week = seq(1, 7, 2), clip = FALSE), "* * * * 0/2")
 expect_equal(r2cron(days_week = c(1, 3), clip = FALSE), "* * * * 0,2")
 expect_equal(r2cron(days_week = c(1, 7), clip = FALSE), "* * * * 0,6")
 expect_equal(r2cron(days_week = c(1, 3, 4), clip = FALSE), "* * * * 0,2,3")

 expect_error(r2cron(days_week = "x", clip = FALSE))
 expect_error(r2cron(days_week = TRUE, clip = FALSE))
 expect_error(r2cron(days_week = mtcars, clip = FALSE))
 expect_error(r2cron(days_week = list("x", "y"), clip = FALSE))
 expect_error(r2cron(days_week = 1.2, clip = FALSE))
 expect_error(r2cron(days_week = 8, clip = FALSE))

})

test_that("r2cron() out-of-range day-month values are warned", {

 expect_warning(r2cron(days_month = 29, months = 2, clip = FALSE))
 expect_warning(r2cron(days_month = 30, months = 2, clip = FALSE))
 expect_warning(r2cron(days_month = 31, months = 2, clip = FALSE))
 expect_warning(r2cron(days_month = 31, months = 4, clip = FALSE))
 expect_warning(r2cron(days_month = 31, months = 6, clip = FALSE))
 expect_warning(r2cron(days_month = 31, months = 9, clip = FALSE))
 expect_warning(r2cron(days_month = 31, months = 11, clip = FALSE))
 expect_warning(r2cron(days_month = 29, months = 1:2, clip = FALSE))
 expect_warning(r2cron(days_month = 31, months = c(1, 4), clip = FALSE))

})

test_that("r2cron() clipboard-copying is supported", {

  msg <- "Copied to clipboard"
  skip_on_os("linux")  # R CMD check fails on Ubuntu because {clipr}
  expect_output(r2cron(clip = TRUE), msg)
  expect_output(r2cron(minutes = 1, clip = TRUE), msg)

})


# cron2eng() --------------------------------------------------------------


test_that("cron2eng() errors with bad input", {

  expect_error(cron2eng(1))
  expect_error(cron2eng("x"))
  expect_error(cron2eng("x x x x x"))
  expect_error(cron2eng("60 * * * *"))
  expect_error(cron2eng("* * * *"))
  expect_error(cron2eng(TRUE))
  expect_error(cron2eng(mtcars))
  expect_error(cron2eng(list("x", "y")))

})

test_that("cron2eng() outputs as expected", {

  expect_output(cron2eng())
  expect_output(cron2eng("1-3 1 1,3,4 1 1"))
  expect_output(cron2eng("* 1/5 * 1/3 0/2"))

})

test_that(".vec2eng() does its job", {

  expect_error(dialga:::.vec2eng(iris))
  expect_error(dialga:::.vec2eng(list(x = 1, y = 1)))
  expect_equal(dialga:::.vec2eng(letters[1]), "a")
  expect_equal(dialga:::.vec2eng(letters[1:2]), "a and b")
  expect_equal(dialga:::.vec2eng(letters[1:3]), "a, b and c")

})


R/utils.R
#' Internal Stopifnot Procedure
#'
#' Check that inputs to each time-period argument of \code{\link{r2cron}} are
#' valid.
#'
#' @param x The user's input for a time period argument in the
#'     \code{\link{r2cron}} function
#' @param p Valid cron time periods used in the \code{\link{r2cron}} function.
#'
#' @return Error if it fails.
#'
.stop <- function(x, p = c("minutes", "hours", "days_month",
                           "months", "days_week")) {

  # Set integer ranges for each valid cron time-period
  r <- if (p == "minutes") {
    0L:59L
  } else if (p == "hours") {
    0L:23L
  } else if (p == "days_month") {
    1L:31L
  } else if (p == "months") {
    1L:12L
  } else if (p == "days_week") {
    1L:7L
  }

  if (is.logical(x) |
      is.character(x) & !all(x == "*") |
      is.numeric(x) & !all(x %in% r)) {

    # Fail if input is outside valid range for the time period
    stop(
      paste0(
        "'", p, "' argument must be integers from ",
        min(r), " to ", max(r)
      )
    )

  }

}

#' Internal R Expression to Cron String Procedure
#'
#' Convert time-period arguments of \code{\link{r2cron}} from R expressions to
#' the most appropriate cron expression format and paste them together.
#'
#' @param x The user's input for a time period argument in the
#'     \code{\link{r2cron}} function
#' @param p Valid cron time periods used in the \code{\link{r2cron}} function.
#'
#' @return A character string.
#'
.as_cron <- function(x, p) {

  if ((p == "minutes"    & length(x) == 60 & all(x %in% 0L:59L)) |
      (p == "hours"      & length(x) == 24 & all(x %in% 0L:23L)) |
      (p == "days_month" & length(x) == 31 & all(x %in% 1L:31L)) |
      (p == "months"     & length(x) == 12 & all(x %in% 1L:12L)) |
      (p == "days_week"  & length(x) == 7  & all(x %in% 0L:6L))) {

    "*"  # i.e. every unit of the time period

  } else if (is.numeric(x) & length(x) == 1) {

    as.character(x)  # i.e. 1 becomes "1"

  } else if (is.numeric(x) & length(x) > 1 & all(diff(x) == 1)) {

    paste0(x[1], "-", x[length(x)])  # i.e. 1:3 becomes "1-3"

  } else if (is.numeric(x) & length(x) > 2 & length(unique(diff(x))) == 1) {

    if (p == "minutes"    & length(x) == length(seq(x[1], 59, unique(diff(x)))) |
        p == "hours"      & length(x) == length(seq(x[1], 23, unique(diff(x)))) |
        p == "days_month" & length(x) == length(seq(x[1], 31, unique(diff(x)))) |
        p == "months"     & length(x) == length(seq(x[1], 12, unique(diff(x)))) |
        p == "days_week"  & length(x) == length(seq(x[1], 6,  unique(diff(x))))) {

      paste0(x[1], "/", unique(diff(x)))  # i.e. seq(0, 59, 20) is "0/20"

    } else {

      paste0(x, collapse = ",")  # i.e. c(1, 2, 5) becomes "1,2,5"

    }

  } else {

    paste0(x, collapse = ",")  # i.e. c(1, 2, 5) becomes "1,2,5"

  }

}

#' Convert a Vector to a Sentence
#'
#' Paste elements of a vector into sentence form, so that items are separated by
#' commas and the last is separated with the word 'and'. Used for output in
#' \code{\link{cron2eng}}.
#'
#' @param x A vector.
#'
#' @return A character string.
#'
.vec2eng <- function(x) {

  if(!is.atomic(x)) {
    stop(".vec2eng input must be a vector")
  }

  if (length(x) == 1) {
    x
  } else if (length(x) > 1) {
    paste(paste(x[1:length(x)-1], collapse = ", "), "and", x[length(x)])
  }

}


R/cron.R
#' Build Cron Strings from R Expressions
#'
#' Provide an R expression to each of the arguments, which are different time
#' periods (minutes, hours, days of the month, months, days of the week), and
#' receive an equivalent cron-expression string in return. See 'arguments' for
#' valid value ranges and 'details' for valid input formats.
#'
#' @param minutes Minutes past the hour, integers 0 to 59 inclusive.
#' @param hours Hours on a 24-hour clock, integers 0 to 23 inclusive.
#' @param days_month Day number of the month, integers 1 to 31 inclusive.
#' @param months Month number, integers 1 to 12 inclusive.
#' @param days_week Day of the week, integers 1 to 7 inclusive (where Sunday is
#'     the first day of the week).
#' @param clip Logical. Copy output to clipboard? Windows, macOS and X11 only.
#'     Requires installation of the {clipr} package.
#'
#' @details The time-period arguments default to every unit within that time
#'     period, like 'every minute, of every hour, of every day of the month,
#'     etc'. The following input formats are acceptable inputs to all the time
#'     period arguments:
#' \itemize{
#'   \item a single integer value, like \code{1}
#'   \item consecutive-integer vectors, like \code{1:3}
#'   \item nonconsecutive, irregularly-spaced integer vectors, like
#'       \code{c(2, 3, 5)}
#'   \item regularly-spaced integer sequences with specified start and end
#'   values, like \code{seq(3, 59, 15)} (useful for specifying sequences within
#'   the full time period,'every 15 minutes of the hour starting from minute 3',
#'   like in this example)
#' }
#'
#' @return A character string (copied to the clipboard at the user's
#'     discretion).
#' @export
#'
#' @examples \dontrun{
#' r2cron(
#'   minutes = seq(0, 59, 20),
#'   hours = 15:17,  # 24-hr clock
#'   days_month = 1,
#'   months = c(4, 10, 11),
#'   days_week = c(1, 7),  # Sunday is '1'
#'   clip = FALSE
#' )}
#'
r2cron <- function(minutes = 0L:59L, hours  = 0L:23L,
                   days_month  = 1L:31L, months = 1L:12L, days_week  = 1L:7L,
                   clip = FALSE) {

 # Stop if input is out of range of time period
 .stop(minutes, "minutes")
 .stop(hours, "hours")
 .stop(days_month, "days_month")
 .stop(months, "months")
 .stop(days_week, "days_week")

 # Zero-index the days of the week
 days_week <- days_week - 1

 # Convert each argument's R expression to equivalent cron expression
 cron_minutes <- .as_cron(minutes, "minutes")
 cron_hours <- .as_cron(hours, "hours")
 cron_days_month <- .as_cron(days_month, "days_month")
 cron_months <- .as_cron(months, "months")
 cron_days_week <- .as_cron(days_week, "days_week")

 # Warn if month and day combinations seem unlikely
 if (cron_months != "*" & cron_days_month != "*" &
   any(months %in% c(2, 4, 6, 9, 11)) & any(days_month == 31)) {
   warning("\n  Sure? There's no 31st in Feb, Apr, Jun, Sept nor Nov.\n")
 }
 if (cron_months != "*" & cron_days_month != "*" &
     any(months == 2) & any(days_month == 30)) {
   warning("\n  Sure? There's no 30th in Feb.\n")
 }
 if (cron_months != "*" & cron_days_month != "*" &
     any(months == 2) & any(days_month == 29)) {
   warning("\n  Sure? 29 Feb is only in leap years.\n")
 }

 # Paste together cron expressions for each time period
 cron_string <- paste(
  cron_minutes, cron_hours, cron_days_month, cron_months, cron_days_week
 )

 # Copy the string to the user's clipboard if requested
 if (clip) {
  clipr::write_clip(cron_string, allow_non_interactive = TRUE)
  cat("Copied to clipboard\n")
 }

 # Return the full cron expression
 return(cron_string)

}

#' Generate English Interpretation of Valid Cron Strings
#'
#' Demystify the meaning of a valid cron string by converting it to an
#' equivalent sentence in English. Can take the output from \code{\link{r2cron}}
#' for example. (Under development.)
#'
#' @param cron Character. A valid cron expression, i.e. a string with five
#'     time period 'slots' (minutes, hours, days of the month, months, days of
#'     the week), separated by spaces. See details for more information.
#'
#' @details The cron string slots and their valid ranges are:
#' \itemize{
#'   \item Slot 1: minutes past the hour, integers 0 to 59 inclusive.
#'   \item Slot 2: hours on a 24-hour clock, integers 0 to 23 inclusive.
#'   \item Slot 3: day number of the month, integers 1 to 31 inclusive.
#'   \item Slot 4: month number, integers 1 to 12 inclusive.
#'   \item Slot 5: day of the week, integers 0 to 6 inclusive (where Sunday is
#'     the first day of the week).
#' }
#'
#' In addition, the following input formats are acceptable to all the time
#' slots:
#' \itemize{
#'   \item a single integer value, like \code{1}
#'   \item consecutive-integer vectors, like \code{1:3}
#'   \item nonconsecutive, irregularly-spaced integer vectors, like
#'       \code{c(2, 3, 5)}
#'   \item regularly-spaced integer sequences with specified start and end
#'       values, like \code{seq(3, 59, 15)} (useful for specifying
#'       sequences within the full time period,'every 15 minutes of the
#'       hour starting from minute 3', like in this example)
#' }
#'
#' @return Result printed to console. An English sentence interpretation of the
#'     cron string that was the input.
#' @export
#'
#' @examples \dontrun{
#' cron2eng("1,2,5 2-3 * 1/3 5")}
#'
cron2eng <- function(cron = "* * * * *") {

  if (!is.character(cron)) {
    stop("Argument 'cron' must be a valid character-class cron expression.")
  }

  if (length(strsplit(cron, " ")[[1]]) != 5) {
    stop("Argument 'cron' must be a valid cron expression.")
  }

  if (length(strsplit(cron, " ")[[1]]) != 5 |  # 5 time slots
      !all(grepl("\\d|\\*|\\s|,|/|-", strsplit(cron, "")[[1]])) |  # valid chars
      !all(grepl("\\b[0-9]\\b|\\b[1-5][0-9]\\b|\\*",
                 strsplit(cron, "\\s|,|/|-")[[1]])) ) {  # valid number range
    stop(
      "Argument 'cron' must have five 'time slots' separated by spaces.\n",
      "Slots must contain valid integers or '*', '/', '-' and ',."
    )
  }

  # Split each time period to a list element
  p_list <- as.list(strsplit(cron, " ")[[1]])

  # Name elements with text to be used in output for that time period
  names(p_list) <- c(
    "minute(s)", "hour(s)", "day(s) of the month",
    "month(s)", "day(s) of the week"
  )

  if (grepl("\\d", p_list[[2]])) {  # if minutes slot contains a number

    if (grepl("/", p_list[[2]])) {  # if minutes slot contains a slash

      # Extract value "x" (start hour) from "x/y"
      slot_val <- as.numeric(strsplit(p_list[[2]], "/")[[1]][1])

      # Replace "x" with corresponding time string (hour)
      if (slot_val == 0) {
        replacement <- "12AM"
      } else if (slot_val >= 1 & slot_val <= 11) {
        replacement <- paste0(slot_val, "AM")
      } else if (slot_val == 12) {
        replacement <- "12PM"
      } else if (slot_val >= 13 & slot_val <= 23) {
        replacement <- paste0(slot_val - 12, "PM")
      }

      # Replace the "x" in "x/y" with time string (hour)
      p_list[[2]] <-
        gsub("^\\d{1,2}(?=/)", replacement, p_list[[2]], perl = TRUE)

    } else {

      # Find/replace any hours in the form "x", "x,y,z", "x-y"
      p_list[[2]] <- gsub("\\b0\\b",  "12AM", p_list[[2]])
      p_list[[2]] <- gsub("\\b1\\b",  "1AM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b2\\b",  "2AM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b3\\b",  "3AM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b4\\b",  "4AM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b5\\b",  "5AM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b6\\b",  "6AM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b7\\b",  "7AM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b8\\b",  "8AM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b9\\b",  "9AM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b10\\b", "10AM", p_list[[2]])
      p_list[[2]] <- gsub("\\b11\\b", "11AM", p_list[[2]])
      p_list[[2]] <- gsub("\\b12\\b", "12PM", p_list[[2]])
      p_list[[2]] <- gsub("\\b13\\b", "1PM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b14\\b", "2PM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b15\\b", "3PM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b16\\b", "4PM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b17\\b", "5PM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b18\\b", "6PM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b19\\b", "7PM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b20\\b", "8PM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b21\\b", "9PM",  p_list[[2]])
      p_list[[2]] <- gsub("\\b22\\b", "10PM", p_list[[2]])
      p_list[[2]] <- gsub("\\b23\\b", "11PM", p_list[[2]])

    }

  }

  if (grepl("\\d", p_list[[4]])) {  # if months slot contains a number

    if (grepl("/", p_list[[4]])) {  # if months slot contains a slash

      # Extract value "x" (start month) from "x/y"
      slot_val <- as.numeric(strsplit(p_list[[4]], "/")[[1]][1])

      # Replace "x" with corresponding time string (month)
      replacement <- month.name[slot_val]

      # Replace the "x" in "x/y" with time string (month)
      p_list[[4]] <-
        gsub("^\\d{1,2}(?=/)", replacement, p_list[[4]], perl = TRUE)

    } else {

      # Find/replace any months in the form "x", "x,y,z", "x-y"
      p_list[[4]] <- gsub("\\b1\\b",  "January",   p_list[[4]])
      p_list[[4]] <- gsub("\\b2\\b",  "February",  p_list[[4]])
      p_list[[4]] <- gsub("\\b3\\b",  "March",     p_list[[4]])
      p_list[[4]] <- gsub("\\b4\\b",  "April",     p_list[[4]])
      p_list[[4]] <- gsub("\\b5\\b",  "May",       p_list[[4]])
      p_list[[4]] <- gsub("\\b6\\b",  "June",      p_list[[4]])
      p_list[[4]] <- gsub("\\b7\\b",  "July",      p_list[[4]])
      p_list[[4]] <- gsub("\\b8\\b",  "August",    p_list[[4]])
      p_list[[4]] <- gsub("\\b9\\b",  "September", p_list[[4]])
      p_list[[4]] <- gsub("\\b10\\b", "October",   p_list[[4]])
      p_list[[4]] <- gsub("\\b11\\b", "November",  p_list[[4]])
      p_list[[4]] <- gsub("\\b12\\b", "December",  p_list[[4]])

    }

  }

  if (grepl("\\d", p_list[[5]])) {

    if (grepl("/", p_list[[5]])) {

      # Extract value "x" (start hour) from "x/y"
      slot_val <- as.numeric(strsplit(p_list[[5]], "/")[[1]][1]) + 1

      # Replace "x" with corresponding time string (day of week)
      replacement <- c("Sunday", "Monday", "Tuesday", "Wednesday",
                       "Thursday", "Friday", "Saturday")[slot_val]

      # Replace the "x" in "x/y" with time string (day of week)
      p_list[[5]] <-
        gsub("\\d{1,2}(?=/)", replacement, p_list[[5]], perl = TRUE)

    } else {

      # Find/replace any days-of-week in the form "x", "x,y,z", "x-y"
      p_list[[5]] <- gsub("0", "Sunday",    p_list[[5]])
      p_list[[5]] <- gsub("1", "Monday",    p_list[[5]])
      p_list[[5]] <- gsub("2", "Tuesday",   p_list[[5]])
      p_list[[5]] <- gsub("3", "Wednesday", p_list[[5]])
      p_list[[5]] <- gsub("4", "Thursday",  p_list[[5]])
      p_list[[5]] <- gsub("5", "Friday",    p_list[[5]])
      p_list[[5]] <- gsub("6", "Saturday",  p_list[[5]])

    }

  }

  # Convert each time slot's cron expression to English
  for (p in names(p_list)) {

    if (p_list[[p]] == "*") {

      # Every valid unit of the time slot
      if (p == names(p_list)[5]) {
        p_list[[p]] <- paste("any", p)  # more specific for day of week
      } else {
        p_list[[p]] <- paste("every", p)
      }

    } else if (grepl("^\\d{1,2}$|^\\w{1,}$", p_list[[p]])) {

      # Single integer value
      p_list[[p]] <- paste(p, p_list[[p]])

    } else if (grepl("^\\w{1,}-\\w{1,}$", p_list[[p]])) {

      # Consecutive units with non-min start and non-max stop value
      n_split <- strsplit(p_list[[p]], "-")
      p_list[[p]] <- paste(
        p, n_split[[1]][1], "to", n_split[[1]][2]
      )

    } else if (grepl("^\\w{1,},", p_list[[p]])) {

      # Nonconsecutive, irregularly spaced integers
      n_units <- strsplit(p_list[[p]], ",")[[1]]
      p_list[[p]] <- paste(p, .vec2eng(n_units))

    } else if (grepl("^\\w{1,}/\\w{1,}$", p_list[[p]])) {

      # regularly-spaced integer sequences with specified start and end
      n_split <- strsplit(p_list[[p]], "/")
      p_list[[p]] <- paste0(
        "every ", n_split[[1]][2], " ", p,
        " starting from ",
        ifelse(p %in% c("minute(s)", "day(s)"), paste0(p, " "), ""),
        n_split[[1]][1]
      )

    }

  }

  # Collate sentences for each time period
  cat(
    paste0("Cron string '", cron, "' means:"),
    paste0("  - ", p_list[[1]]),
    paste0("  - ", p_list[[2]]),
    paste0("  - ", p_list[[3]]),
    paste0("  - ", p_list[[4]]),
    ifelse(  # a clear 'and' for user-specified days of the week
      grepl("^day", p_list[[5]]),
      paste0("  - and ", p_list[[5]]),
      paste0("  - ", p_list[[5]])
    ),
    sep = "\n"
  )

}
